Parser/tokenizer.c

```diff
@@ -489,25 +489,37 @@ static void fp_ungetc(int c, struct tok_state *tok) {
 
 /* Check whether the characters at s start a valid
    UTF-8 sequence. Return the number of characters forming
-   the sequence if yes, 0 if not.  */
+   the sequence if yes, 0 if not.  The special cases match
+   those in stringlib/codecs.h:decode_utf8.
+*/
 static int valid_utf8(const unsigned char* s)
 {
     int expected = 0;
     int length;
-    if (*s < 0x80)
+    if (*s < 0x80) {
         /* single-byte code */
         return 1;
-    if (*s < 0xc0)
-        /* following byte */
-        return 0;
-    if (*s < 0xE0)
+    } else if (*s < 0xE0) {
+        if (*s < 0xC2) {
```

Comment:
Could you repeat the comments from stringlib/codecs.h here? Just to make it easier to understand the code here (not all of it is structured exactly as the code there). E.g. in this case
```
            /* \xC2\x80-\xDF\xBF -- 0080-07FF */
```
Also, I take it the `IS_CONTINUATION_BYTE()` checks from there are not needed here? (Why not?)

Reply:
+1 on the comments.

The `IS_CONTINUATION_BYTE` checks happen [further down in this function](https://github.com/python/cpython/blob/main/Parser/tokenizer.c#L512) in the loop.  `stringlib/codecs.h:utf8_decode` does those checks inline with each different size, whereas this function sets the size with `expected` and then does a loop over that later.  There might be some maintenance advantage to keeping the code here and in `utf8_decode` following the same pattern, given that it wasn't obvious how they are equivalent.  Do you think that's worth changing here as well?